compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala

```diff
@@ -60,35 +60,45 @@ object JavaNullInterop {
     if (sym.name == nme.TYPE_ || sym.isAllOf(Flags.JavaEnumValue))
       // Don't nullify the `TYPE` field in every class and Java enum instances
       tp
-    else if (sym.name == nme.toString_ || sym.isConstructor)
-      // Don't nullify the return type of the `toString` method and constructors
-      nullifyParamsOnly(tp)
+    else if (sym.name == nme.toString_ || sym.isConstructor || hasNotNullAnnot(sym))
+      // Don't nullify the return type of the `toString` method.
+      // Don't nullify the return type of constructors.
+      // Don't nullify the return type of methods with a not-null annotation.
+      nullifyExceptReturnType(tp)
     else
       // Otherwise, nullify everything
       nullifyType(tp)
   }
 
-  /** Only nullify method parameters (but not result types). */
-  private def nullifyParamsOnly(tp: Type)(implicit ctx: Context): Type =
-    new JavaNullMap(alreadyNullable = false, nonNullResultType = true)(ctx)(tp)
+  private def hasNotNullAnnot(sym: Symbol)(implicit ctx: Context): Boolean =
+    ctx.definitions.NotNullAnnots.exists(nna => sym.unforcedAnnotation(nna).isDefined)
+
+  /** If tp is a MethodType, the parameters and the inside of return type are nullified,
+   *  but the result return type is not nullable.
+   *  If tp is a type of a field, the inside of the type is nullified,
+   *  but the result type is not nullable.
+  */
+  private def nullifyExceptReturnType(tp: Type)(implicit ctx: Context): Type =
+    new JavaNullMap(true)(ctx)(tp)
 
   /** Nullifies a Java type by adding `| JavaNull` in the relevant places. */
   private def nullifyType(tp: Type)(implicit ctx: Context): Type =
-    new JavaNullMap(alreadyNullable = false, nonNullResultType = false)(ctx)(tp)
+    new JavaNullMap(false)(ctx)(tp)
 
 
   /** A type map that adds `| JavaNull`.
-   *  @param alreadyNullable whether the type being mapped is already nullable (at the outermost level).
-   *                         This is needed so that `JavaNullMap(A | B)` gives back `(A | B) | JavaNull`,
-   *                         instead of `(A|JavaNull | B|JavaNull) | JavaNull`.
-   *  @param nonNullResultType if true, then the current type is a method or generic method type,
-   *                           and we don't want to nullify its return type at the top level.
-   *                           This is useful e.g. for constructors.
+   *  If a field has a final flag and its value is not null, then the type of it is a ConstantType,
```

Comment:
This comment belongs below line 151:

// In all other cases, return the type unchanged.
// In particular, if the type is a ConstantType, then we don't nullify it because it is the type of a final non-nullable field.
case _ => tp

Reply:
Comment added